src/diagnostics/MOM_wave_structure.F90

```diff
@@ -409,78 +417,85 @@ subroutine wave_structure(h, tv, G, GV, US, cn, ModeNum, freq, CS, En, full_halo
+            b_dom(:) = 0.0
             c_diag(:) = 0.0
             e_guess(:) = 0.0
             e_itt(:) = 0.0
             w_strct(:) = 0.0
             do K=3,kc-1
               row = K-1 ! indexing for TD matrix rows
-              gp_unscaled = gprime(K)
-              lam_z(row) = lam*gp_unscaled
-              a_diag(row) = gp_unscaled*(-Igu(K))
-              b_diag(row) = gp_unscaled*(Igu(K)+Igl(K)) - lam_z(row)
-              c_diag(row) = gp_unscaled*(-Igl(K))
+              lam_z(row) = lam*gprime(K)
+              a_diag(row) = gprime(K)*(-Igu(K))
+              b_dom(row) = 2.0*gprime(K)*(Igu(K)+Igl(K)) - lam_z(row)
+              c_diag(row) = gprime(K)*(-Igl(K))
+            enddo
+            if (CS%debug) then ; do row=2,kc-2
               if (isnan(lam_z(row)))then  ; print *, "Wave_structure: lam_z(row) is NAN" ; endif
               if (isnan(a_diag(row)))then ; print *, "Wave_structure: a(k) is NAN" ; endif
-              if (isnan(b_diag(row)))then ; print *, "Wave_structure: b(k) is NAN" ; endif
               if (isnan(c_diag(row)))then ; print *, "Wave_structure: c(k) is NAN" ; endif
-            enddo
+            enddo ; endif
             ! Populate top row of tridiagonal matrix
             K=2 ; row = K-1 ;
-            gp_unscaled = gprime(K)
-            lam_z(row) = lam*gp_unscaled
+            lam_z(row) = lam*gprime(K)
             a_diag(row) = 0.0
-            b_diag(row) = gp_unscaled*(Igu(K)+Igl(K)) - lam_z(row)
-            c_diag(row) = gp_unscaled*(-Igl(K))
+            b_dom(row) = gprime(K)*(Igu(K)+2.0*Igl(K)) - lam_z(row)
+            c_diag(row) = gprime(K)*(-Igl(K))
             ! Populate bottom row of tridiagonal matrix
             K=kc ; row = K-1
-            gp_unscaled = gprime(K)
-            lam_z(row) = lam*gp_unscaled
-            a_diag(row) = gp_unscaled*(-Igu(K))
-            b_diag(row) = gp_unscaled*(Igu(K)+Igl(K)) - lam_z(row)
+            lam_z(row) = lam*gprime(K)
+            a_diag(row) = gprime(K)*(-Igu(K))
+            b_dom(row) = gprime(K)*(2.0*Igu(K) + Igl(K)) - lam_z(row)
             c_diag(row) = 0.0
 
-            ! Guess a vector shape to start with (excludes surface and bottom)
-            e_guess(1:kc-1) = sin((z_int(2:kc)/htot(i,j)) *Pi)
-            e_guess(1:kc-1) = e_guess(1:kc-1)/sqrt(sum(e_guess(1:kc-1)**2))
+            ! Guess a normalized vector shape to start with (excludes surface and bottom)
+            emag2 = 0.0
+            pi_htot = Pi / htot(i,j)
+            do K=2,kc
+              e_guess(K-1) = sin(pi_htot * z_int(K))
+              emag2 = emag2 + e_guess(K-1)**2
+            enddo
+            renorm = 1.0 / sqrt(emag2)
+            do K=2,kc ; e_guess(K-1) = renorm*e_guess(K-1) ; enddo
 
             ! Perform inverse iteration with tri-diag solver
             do itt=1,max_itt
               ! this solver becomes unstable very quickly
+              ! b_diag(1:kc-1) = b_dom(1:kc-1) + (a_diag(1:kc-1) + c_diag(1:kc-1))
```

Comment:
I think this should be

```fortran
! b_diag(1:kc-1) = b_dom(1:kc-1) - (a_diag(1:kc-1) + c_diag(1:kc-1))
```

Reply:
Yes, you are correct about the sign error in the comment, Raf.  This sign error has now been corrected.